hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientResponseHandlerSupplier.java

```diff
@@ -0,0 +1,205 @@
+            handleClientMessage(message);
+        } catch (Exception e) {
+            logger.severe("Failed to process task: " + new ClientPacket(connection, message)
+                    + " on responseThread: " + Thread.currentThread().getName(), e);
+        } finally {
+            connection.decrementPendingPacketCount();
+        }
+    }
+
+    private void handleClientMessage(ClientMessage clientMessage) {
+        long correlationId = clientMessage.getCorrelationId();
+
+        ClientInvocation future = invocationService.deRegisterCallId(correlationId);
+        if (future == null) {
+            logger.warning("No call for callId: " + correlationId + ", response: " + clientMessage);
+            return;
+        }
+
+        if (ErrorCodec.TYPE == clientMessage.getMessageType()) {
+            future.notifyException(client.getClientExceptionFactory().createException(clientMessage));
+        } else {
+            future.notify(clientMessage);
+        }
+    }
+
+    private class ResponseThread extends Thread {
+        private final BlockingQueue<ClientPacket> responseQueue;
+
+        ResponseThread(String name) {
+            super(name);
+            setContextClassLoader(client.getClientConfig().getClassLoader());
+            this.responseQueue = new MPSCQueue<ClientPacket>(this, getIdleStrategy(client.getProperties(), IDLE_STRATEGY));
+        }
+
+        @Override
+        public void run() {
+            try {
+                doRun();
+            } catch (OutOfMemoryError e) {
+                onOutOfMemory(e);
+            } catch (Throwable t) {
+                invocationService.invocationLogger.severe(t);
+            }
+        }
+
+        private void doRun() {
+            while (!invocationService.isShutdown()) {
+                ClientPacket task;
+                try {
+                    task = responseQueue.take();
+                } catch (InterruptedException e) {
+                    continue;
+                }
+                process(task.connection, task.message);
+            }
+        }
+    }
+
+    private class SyncResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            process(connection, message);
+        }
+    }
+
+    private class AsyncSingleThreadedResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            responseThreads[0].responseQueue.add(new ClientPacket(connection, message));
+        }
+    }
+
+    private class AsyncMultiThreadedResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            int threadIndex = INT_HOLDER.get().getAndInc() % responseThreads.length;
+            responseThreads[threadIndex].responseQueue.add(new ClientPacket(connection, message));
```

Comment:
Will this cause ordering problems ? 
Ordering is important for events. I understand that this change will not affect how events are handled. So question is:

Do we need any ordering for responses ?
Do we need to put the response messages from same partition id to same thread ?
```
int partitionId = message.getPartitionId();
int threadIndex = (partitionId == -1 ? INT_HOLDER.get().getAndInc() : partitionId) % responseThreads.length;
```

Even tough I am asking this, I could not really find a use-case where ordering of responses is important. Only thing comes to mind is `andThen` calls with `async` methods. Although configurable, there are 3 threads by default already(order is no preserved anyway) 

So I am just asking if this issue is considered.

Reply:
Ordering for responses is not relevant (if it is, somebody did something very stupid).

Ordering is determines by invocation start/completion. If 2 operations are executed concurrently, any order is fine. So therefor any order of responses is fine as well.